docs/genesis-preparation.md

```diff
@@ -0,0 +1,352 @@
+
+```
+$ stack exec -- cardano-keygen generate-key --path /tmp/node1.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node2.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node3.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node4.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node5.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node6.key
+$ stack exec -- cardano-keygen generate-key --path /tmp/node7.key
+```
+
+**Responsible**: The secret key for particular node should generate a person who
+launch this node. For Mainnet launching secret keys generation should be done by
+IOHK DevOps team.
+
+### 1.2. Public Keys
+
+It's possible to read generated secret key by this command:
+
+```
+$ stack exec -- cardano-keygen read-key --path /tmp/node1.key
+```
+
+Example of response:
+
+```
+[keygen:INFO] Processing command
+[keygen:INFO] Primary: 9645e94f800a5dbcf2ea4f21c60755ca4b75ba7efe02ea20a8278d1e2ff185f1c22cd797a8c5e131a40906f89d57fcaea6c70dcdf3869e4a6a7a6a04292ba10f; address hash: 204779e68379f19ad88c3599c22015988febb4c0258f100a61346ce1
+[keygen:INFO] No wallet set
+[keygen:INFO] Keys:
+[keygen:INFO] Vss PK: 582103803acf1550b381926d748949d5bd4d8a7b7cfe1f8112f52555d49ff88ac6db77; short = vsspub:6cc4ca60
+```
+
+`Primary` line contains public key. So all public keys should be extracted from the
+secret ones and published in any public channel: GitHub Gist, website, etc. Address hash
+is a stakeholder's identifier which will be included in genesis data as well (please see
+"Genesis JSON Format" below).
+
+**Responsible**: The secret key for particular node should generate a person who
+launch this node. For Mainnet launching public keys publishing should be done by
+IOHK DevOps team.
+
+## 2. Bootstrap Stakeholders Keys
+
+### 2.1. Secret Keys 
+
+Each of the parties (IOHK/CGG/CF) generates its secret/public keys using the same command:
+
+```
+$ stack exec -- cardano-keygen generate-key --path PATH
+```
+
+Please note that purpose of these keys is **different** than keys generated in step `1.1`.
+These keys will be used for heavyweight delegation (see the next step).
+
+**Responsible**: Each of the parties' (IOHK/CGG/CF) DevOps teams.
+
+### 2.2. Public Keys
+
+Corresponding public keys can be obtained from generated secret keys the same way as in step
+`1.2`.
+
+**Responsible**: Each of the parties' (IOHK/CGG/CF) DevOps teams.
+
+### 2.3. Delegation Certificates
+
+Each of the parties (IOHK/CGG/CF) generates delegation certificate for the nodes. This
+certificate allows to delegate stake from `SK` to `DPK`, where `SK` is a stakeholder's
+secret key and `DPK` is a delegate's public key. Stakeholder's secret key was generated
+in step `2.1`, and delegate's public key was obtained in step `1.2`.
+
+Firts of all, you should add the key by this command:
+
+```
+$ stack exec -- cardano-auxx cmd --commands add-key PATH_TO_SK 
```

Comment:
This doesn't work as described.

```
[jake@godel:~/src/cardano-sl/auxx]$ stack exec --nix -- cardano-auxx cmd --commands add-key "../keys/stakeholder.key"
trace: fetching nixpkgs
Invalid argument `../keys/stakeholder.key'

Usage: cardano-auxx [--version] COMMAND [--db-path FILEPATH] [--rebuild-db]
                    [--spending-genesis INT] [--vss-genesis INT]
                    [--keyfile FILEPATH] [--backup-phrase PHRASE]
                    [--topology FILEPATH] [--kademlia FILEPATH]
                    [--node-id NODE_ID] [--default-port PORT]
                    [--policies FILEPATH] [--address IP:PORT] [--listen IP:PORT]
                    [--json-log FILEPATH] [--log-config FILEPATH]
                    [--logs-prefix FILEPATH] [--report-server URI]
                    [--update-server URI] [--configuration-file FILEPATH]
                    [--configuration-key TEXT] [--system-start TIMESTAMP]
                    [--update-latest-path FILEPATH] [--update-with-package]
                    [--no-ntp] [--metrics] [--ekg-server IP:PORT]
                    [--statsd-server IP:PORT] [--statsd-interval MILLISECONDS]
                    [--statsd-debug BOOL] [--statsd-prefix TEXT]
                    [--statsd-suffix TEXT] [--dump-genesis-data-to ARG]
                    [--peer HOST:PORT]
  Cardano SL CLI utilities.
```

Reply:
Try
```
[jake@godel:~/src/cardano-sl/auxx]$ stack exec --nix -- cardano-auxx cmd --commands "add-key ../keys/stakeholder.key"
```

Reply:
I tried that, and then based on the error added `--start-time 0` as well.

Now I get:

```
+ stack exec --nix -- cardano-auxx cmd --commands 'add-key ../keys/stakeholder.key' --system-start 0
trace: fetching nixpkgs
ConfigurationOptions {cfoFilePath = "node/configuration.yaml", cfoKey = "default", cfoSystemStart = Just 0}
cardano-auxx: ConfigurationParseFailure "node/configuration.yaml" (InvalidYaml (Just (YamlException "Yaml file not found: node/configuration.yaml")))
```

Reply:
Huh, "file not found". Weird. It's there. At first I figured the file was simply invalid, but now I'm not so sure.

Reply:
@jmitchell try `stack exec -- cardano-auxx --configuration-file node/configuration.mainnet.yaml --configuration-key testnet_public_full --system-start 0 cmd --commands 'listaddr'`

Reply:
Oh, since the `cardano-auxx` isn't build from the top-level dir I've been running it from `./auxx`.

Reply:
Okay, `--configuration-file ../node/configuration.mainnet.yaml` did the trick.